fs/fs.go

```diff
@@ -29,7 +29,7 @@ func NewDatastore(path string) (ds.Datastore, error) {
 
 // KeyFilename returns the filename associated with `key`
 func (d *Datastore) KeyFilename(key ds.Key) string {
-	return filepath.Join(d.path, key.String(), ObjectKeySuffix)
```

Comment:
> The real fix is encoding every segment, e.g. hex. Then the extension being non-hex makes namespace collisions impossible. Then dirs don't need an extension.
> 
> Keys can also be raw binary, which makes the above necessary anyway.

Yeah that seems a good way to do it.

> And the encoding can't be base58 if you care about HFS.

Yep, https://github.com/ipfs/specs/tree/master/repo/fs-repo#blocks-with-an-fs-datastore <-- notes this. we dont have to use the approach listed there-- we can modify it if desired. Just want to keep the objects inspectable by users, like git.

> But if it doesn't look pretty in find, what's the purpose of preserving the key segments, and not just flattening it (treating the key as opaque binary)? To support faster Query prefixes, a feature that isn't used anyway?

Both Query prefixes and user readability.

>  I'm tempted to make the blockstore use a flat version and not this. It's just way simpler and robust, don't have to worry about all the edge cases.

it'll be way easier for us to implement an `s3-datastore` if we keep it in terms of datastore :/ (and even tier them to treat the `fs-datastore` as a local cache).

---

Actually how about this. This fs-datastore is tuned for user viewing-- i.e. the keys are the same -- which is not the best for the go-ipfs repo. Maybe we can make a different fs-datastore that encodes all the keys with hex, etc.